miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/storage/MiniAppSecureStorage.kt

```diff
@@ -0,0 +1,257 @@
+        return securedStorageFile.exists()
+    }
+
+    @Suppress("MagicNumber")
+    fun secureStorageSize(
+        miniAppId: String,
+        onSuccess: (Double) -> Unit,
+        onFailed: (MiniAppStorageError) -> Unit
+    ) {
+        if (isStorageAvailable(miniAppId)) {
+            val sizeInKb = File(secureStorageBasePath, "$miniAppId.txt").length() / (1024.0)
+            onSuccess(sizeInKb)
+        } else {
+            onFailed(MiniAppStorageError.unavailableStorage)
+        }
+    }
+
+    @Suppress("SwallowedException", "TooGenericExceptionCaught")
+    private fun writeToEncryptedFile(
+        miniAppId: String,
+        content: String,
+        onSuccess: (Map<String, String>) -> Unit,
+        onFailed: (MiniAppStorageError) -> Unit
+    ) = try {
+        makeDirectoryAvailable()
+        val fileToWrite = File(secureStorageBasePath, "$miniAppId.txt")
+        if (fileToWrite.exists()) {
+            fileToWrite.delete()
+        }
+        val encryptedFile = EncryptedFile.Builder(
+            activity,
+            fileToWrite,
+            MasterKey.Builder(activity).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build(),
+            EncryptedFile.FileEncryptionScheme.AES256_GCM_HKDF_4KB
+        ).build()
+
+        val fileContent = content.toByteArray(StandardCharsets.UTF_8)
+        encryptedFile.openFileOutput().apply {
+            write(fileContent)
+            flush()
+            close()
+        }
+        onSuccess(deserializeItems(content))
+    } catch (e: Exception) {
+        onFailed(MiniAppStorageError.ioError)
+    }
+
+    @VisibleForTesting
+    @Suppress("SwallowedException", "TooGenericExceptionCaught", "ReturnCount")
+    private fun readFromEncryptedFile(miniAppId: String): Map<String, String>? {
+        try {
+            if (isStorageAvailable(miniAppId)) {
+                val encryptedFile = EncryptedFile.Builder(
+                    activity,
+                    File(secureStorageBasePath, "$miniAppId.txt"),
+                    MasterKey.Builder(activity).setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
+                        .build(),
+                    EncryptedFile.FileEncryptionScheme.AES256_GCM_HKDF_4KB
+                ).build()
+
+                val inputStream = encryptedFile.openFileInput()
+                val plaintext: ByteArray = inputStream.readBytes()
+                val jsonToRead = plaintext.toString(Charsets.UTF_8)
+                return deserializeItems(jsonToRead)
+            } else {
+                return null
+            }
+        } catch (e: Exception) {
+            return null
+        }
+    }
+
+    fun delete(
+        miniAppId: String,
+        onSuccess: () -> Unit,
+        onFailed: (MiniAppStorageError) -> Unit
+    ) {
+        if (isStorageAvailable(miniAppId)) {
+            scope.launch {
+                val file = File(secureStorageBasePath, "$miniAppId.txt")
+                file.delete()
```

Comment:
Could deleting the file have any consequences if the mini app tries to use secure storage again later?
For example, probably when they call `MiniApp.clear()`, they'll then call `MiniApp.setItems()` soon afterwards to set new data.

Could we instead replace the file contents with an empty map - basically just re-initialize the file?

Reply:
@corycaywood we always check the storage is available or not before performing any operation, so if it can not find the storage it just return an error or create a new one if `miniapp` wants to set a new item.